### PR TITLE
[connector/count] Add OTTL conditions to count connector

### PR DESCRIPTION
**Pull request by djaglowski**
This is a major overhaul of the connector to add OTTL conditions.

Notably, this adds support for all OTTL contexts, which means we can now count span events and metrics, in addition to spans, data points, and logs.

Each of the five contexts is a top level config option. Multiple metrics can be defined for each context. Each metric may have zero or more conditionals applied.

### CODE REVIEW

File: connector/countconnector/factory.go [hunk 1]
```diff
@@ -40,20 +47,7 @@ func NewFactory() connector.Factory {
...
-		Metrics: MetricInfo{
-			Name:        defaultMetricNameDataPoints,
-			Description: defaultMetricDescDataPoints,
-		},
-		Logs: MetricInfo{
-			Name:        defaultMetricNameLogRecords,
-			Description: defaultMetricDescLogRecords,
-		},
-	}
+	return &Config{}
```
Comment (djaglowski) [inline connector/countconnector/factory.go:50]: Now using a custom unmarshaler to set defaults so that the default metric for each type can be disabled.

**Comment by djaglowski** [comment]
@TylerHelmuth, would you mind giving the ottl related parts a look over?

**Review by TylerHelmuth — COMMENTED** [review]
OTTL stuff looks correct.  Sorry for the annoying verbosity.  When I can find time I'll try to simplify some of the boilerplate into `internal/filter`

File: connector/countconnector/README.md [hunk 2]
```diff
@@ -29,20 +41,51 @@ exporters:
...
+
+Optionally, emit custom counts by defining metrics under one or more of the following sections:
+
+- `spans`
+- `spanevents`
+-`metrics`
+-`datapoints`
+-`logs`
+
+Under each custom metric name, specify the conditions under which data should be counted.
```
[the remark this answers is not in the document]
Comment (djaglowski) [reply connector/countconnector/README.md:65]: Good point. What's the best doc to link to for an explanation of the syntax?

[the remark this answers is not in the document]
Comment (TylerHelmuth) [reply connector/countconnector/README.md:65]: OTTL's readme is what we link to in other components: https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/pkg/ottl

**Review by TylerHelmuth — APPROVED** [review]
OTTL stuff looks good to me

**Review by kovrus — APPROVED** [review]

**Review by mx-psi — APPROVED** [review]

**Pull request merged**